code/modules/goonchat/browserOutput.dm

```diff
@@ -0,0 +1,289 @@
+	set waitfor = FALSE
+
+	while (owner)
+		ehjax_send(data = owner.is_afk(29) ? "softPang" : "pang") // SoftPang isn't handled anywhere but it'll always reset the opts.lastPang.
+		sleep(30)
+
+/datum/chatOutput/proc/ehjax_send(client/C = owner, window = "browseroutput", data)
+	if(islist(data))
+		data = json_encode(data)
+	C << output("[data]", "[window]:ehjaxCallback")
+
+//Sends client connection details to the chat to handle and save
+/datum/chatOutput/proc/sendClientData()
+	//Get dem deets
+	var/list/deets = list("clientData" = list())
+	deets["clientData"]["ckey"] = owner.ckey
+	deets["clientData"]["ip"] = owner.address
+	deets["clientData"]["compid"] = owner.computer_id
+	var/data = json_encode(deets)
+	ehjax_send(data = data)
+
+//Called by client, sent data to investigate (cookie history so far)
+/datum/chatOutput/proc/analyzeClientData(cookie = "")
+	if(!cookie)
+		return
+
+	if(cookie != "none")
+		var/list/connData = json_decode(cookie)
+		if (connData && islist(connData) && connData.len > 0 && connData["connData"])
+			connectionHistory = connData["connData"] //lol fuck
+			var/list/found = new()
+			for(var/i in connectionHistory.len to 1 step -1)
+				var/list/row = src.connectionHistory[i]
+				if (!row || row.len < 3 || (!row["ckey"] && !row["compid"] && !row["ip"])) //Passed malformed history object
+					return
+				if (world.IsBanned(row["ckey"], row["compid"], row["ip"]))
+					found = row
+					break
+
+			//Uh oh this fucker has a history of playing on a banned account!!
+			if (found.len > 0)
+				//TODO: add a new evasion ban for the CURRENT client details, using the matched row details
+				message_admins("[key_name(src.owner)] has a cookie from a banned account! (Matched: [found["ckey"]], [found["ip"]], [found["compid"]])")
+				log_admin("[key_name(src.owner)] has a cookie from a banned account! (Matched: [found["ckey"]], [found["ip"]], [found["compid"]])")
+
+	cookieSent = TRUE
+
+//Called by js client every 60 seconds
+/datum/chatOutput/proc/ping()
+	return "pong"
+
+//Called by js client on js error
+/datum/chatOutput/proc/debug(error)
+	log_world("\[[time2text(world.realtime, "YYYY-MM-DD hh:mm:ss")]\] Client: [(src.owner.key ? src.owner.key : src.owner)] triggered JS error: [error]")
+
+#ifdef TESTING
+/client/verb/debug_chat()
+	set hidden = TRUE
+	chatOutput.ehjax_send(data = list("firebug" = TRUE))
+#endif
+//Global chat procs
+
+GLOBAL_LIST_EMPTY(bicon_cache)
+
+//Converts an icon to base64. Operates by putting the icon in the iconCache savefile,
+// exporting it as text, and then parsing the base64 from that.
+// (This relies on byond automatically storing icons in savefiles as base64)
+/proc/icon2base64(icon/icon, iconKey = "misc")
+	if (!isicon(icon))
+		return FALSE
+	GLOB.iconCache[iconKey] << icon
+	var/iconData = GLOB.iconCache.ExportText(iconKey)
+	var/list/partial = splittext(iconData, "{")
+	return replacetext(copytext(partial[2], 3, -5), "\n", "")
+
+/proc/bicon(obj)
+	if (!obj)
```

Comment:
IMO make these take atom/A as argument and then adjust their code to account for that